crates/notion-core/src/distro/package.rs

```diff
@@ -379,28 +402,83 @@ impl PackageVersion {
         bin_name: String,
         bin_path: String,
         platform_spec: &PlatformSpec,
+        loader: Option<BinLoader>,
     ) -> BinConfig {
         BinConfig {
             name: bin_name,
             package: self.name.to_string(),
             version: self.version.clone(),
             path: bin_path,
             platform: platform_spec.clone(),
+            loader,
         }
     }
 
     fn write_config_and_shims(&self, platform_spec: &PlatformSpec) -> Fallible<()> {
         self.package_config(&platform_spec).to_serial().write()?;
         for (bin_name, bin_path) in self.bins.iter() {
-            self.bin_config(bin_name.to_string(), bin_path.to_string(), &platform_spec)
-                .to_serial()
-                .write()?;
+            let loader = self.determine_script_loader(bin_name, bin_path)?;
+            self.bin_config(
+                bin_name.to_string(),
+                bin_path.to_string(),
+                &platform_spec,
+                loader,
+            )
+            .to_serial()
+            .write()?;
             // create a link to the shim executable
             shim::create(&bin_name)?;
         }
         Ok(())
     }
 
+    /// On Unix, shebang loaders work correctly, so we don't need to bother storing loader information
+    #[cfg(unix)]
+    fn determine_script_loader(
+        &self,
+        _bin_name: &str,
+        _bin_path: &str,
+    ) -> Fallible<Option<BinLoader>> {
+        Ok(None)
+    }
+
+    /// On Windows, we need to read the executable and try to find a shebang loader
+    /// If it exists, we store the loader in the BinConfig so that the shim can execute it correctly
+    #[cfg(windows)]
+    fn determine_script_loader(
+        &self,
+        bin_name: &str,
+        bin_path: &str,
+    ) -> Fallible<Option<BinLoader>> {
+        let full_path = bin_full_path(&self.name, &self.version, bin_path, |_| {
+            ErrorDetails::DetermineBinaryLoaderError {
+                bin: bin_name.to_string(),
+            }
+        })?;
+        let script =
+            File::open(full_path).with_context(|_| ErrorDetails::DetermineBinaryLoaderError {
+                bin: bin_name.to_string(),
+            })?;
+        if let Some(Ok(first_line)) = BufReader::new(script).lines().next() {
+            // Note: Regex adapted from @zkochan/cmd-shim package used by Yarn
+            // https://github.com/pnpm/cmd-shim/blob/bac160cc554e5157e4c5f5e595af30740be3519a/index.js#L42
+            let re = Regex::new(r#"^#!\s*(?:/usr/bin/env)?\s*(?P<exe>[^ \t]+) ?(?P<args>.*)$"#)
+                .expect("Regex is valid");
+            if let Some(caps) = re.captures(&first_line) {
+                let args = caps["args"]
```

Comment:
I believe we are confident the groups will always exist, as they are defined in the regex, and we are checking one line above if the regex matches. If there is a match, then each of the groups is necessarily there, since neither of them is defined as optional (though the last could certainly be empty.

Reply:
Ah, I see, it'll be an empty string but always present. Thank you!